File: app/src/main/java/org/lineageos/updater/UpdatesActivity.java
/*
 * Copyright (C) 2017-2026 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.updater;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.icu.text.DateFormat;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SimpleItemAnimator;

import org.lineageos.updater.controller.UpdaterController;
import org.lineageos.updater.controller.UpdaterService;
import org.lineageos.updater.data.Update;
import org.lineageos.updater.download.DownloadClient;
import org.lineageos.updater.misc.Constants;
import org.lineageos.updater.misc.Constants.CheckInterval;
import org.lineageos.updater.misc.DeviceInfoUtils;
import org.lineageos.updater.misc.StringGenerator;
import org.lineageos.updater.misc.Utils;
import org.lineageos.updater.viewmodel.FetchResult;
import org.lineageos.updater.viewmodel.UpdaterViewModel;
import org.lineageos.updater.worker.UpdateCheckWorker;

import java.util.List;

public class UpdatesActivity extends UpdaterBaseActivity implements
        UpdateImporter.Callbacks {

    private static final String TAG = "UpdatesActivity";
    private UpdaterService mUpdaterService;
    private BroadcastReceiver mBroadcastReceiver;

    private UpdatesListAdapter mAdapter;

    private UpdaterViewModel mViewModel;

    private Update mToBeExported = null;
    private final ActivityResultLauncher<Intent> mExportUpdate = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK) {
                    Intent intent = result.getData();
                    if (intent != null) {
                        Uri uri = intent.getData();
                        exportUpdate(uri);
                    }
                }
            });

    private UpdateImporter mUpdateImporter;
    private AlertDialog importDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_updates);

        mViewModel = new ViewModelProvider(this).get(UpdaterViewModel.class);
        mViewModel.getUiStateLiveData().observe(this, state -> {
            boolean hasUpdates = !state.getUpdates().isEmpty();
            if (hasUpdates) {
                findViewById(R.id.no_new_updates_view).setVisibility(View.GONE);
                findViewById(R.id.recycler_view).setVisibility(View.VISIBLE);
            } else {
                findViewById(R.id.no_new_updates_view).setVisibility(View.VISIBLE);
                findViewById(R.id.recycler_view).setVisibility(View.GONE);
            }
            mAdapter.setData(state.getUpdateIds());
            mAdapter.notifyDataSetChanged();
        });
        mViewModel.getUpdateCheckStateLiveData().observe(this, checkState -> {
            setRefreshEnabled(!checkState.isRefreshing());

            if (checkState.getFetchResult() != null) {
                FetchResult result = checkState.getFetchResult();
                if (result instanceof FetchResult.Success) {
                    updateLastCheckedString(checkState.getLastCheckTimestamp());
                    boolean hasNew = ((FetchResult.Success) result).getHasNewUpdates();
                    showToast(hasNew ? R.string.snack_updates_found : R.string.snack_no_updates_found, Toast.LENGTH_SHORT);
                } else if (result instanceof FetchResult.Error) {
                    showToast(R.string.snack_updates_check_failed, Toast.LENGTH_LONG);
                }
                mViewModel.consumeFetchResult();
            }
        });

        mUpdateImporter = new UpdateImporter(this, this);

        RecyclerView recyclerView = findViewById(R.id.recycler_view);
        mAdapter = new UpdatesListAdapter(this, this::exportUpdate);
        recyclerView.setAdapter(mAdapter);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        RecyclerView.ItemAnimator animator = recyclerView.getItemAnimator();
        if (animator instanceof SimpleItemAnimator) {
            ((SimpleItemAnimator) animator).setSupportsChangeAnimations(false);
        }

        mBroadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                if (UpdaterController.ACTION_UPDATE_STATUS.equals(intent.getAction())) {
                    String downloadId = intent.getStringExtra(UpdaterController.EXTRA_DOWNLOAD_ID);
                    handleDownloadStatusChange(downloadId);
                    mAdapter.notifyItemChanged(downloadId);
                } else if (UpdaterController.ACTION_DOWNLOAD_PROGRESS.equals(intent.getAction()) ||
                        UpdaterController.ACTION_INSTALL_PROGRESS.equals(intent.getAction())) {
                    String downloadId = intent.getStringExtra(UpdaterController.EXTRA_DOWNLOAD_ID);
                    mAdapter.notifyItemChanged(downloadId);
                } else if (UpdaterController.ACTION_UPDATE_REMOVED.equals(intent.getAction())) {
                    String downloadId = intent.getStringExtra(UpdaterController.EXTRA_DOWNLOAD_ID);
                    mAdapter.removeItem(downloadId);
                    List<Update> sortedUpdates =
                            mUpdaterService.getUpdaterController().getUpdates();
                    if (sortedUpdates.isEmpty()) {
                        findViewById(R.id.no_new_updates_view).setVisibility(View.VISIBLE);
                        findViewById(R.id.recycler_view).setVisibility(View.GONE);
                    }
                }
            }
        };

        TextView headerTitle = findViewById(R.id.header_title);
        headerTitle.setText(getString(R.string.header_title_text,
                DeviceInfoUtils.getBuildVersion()));

        TextView headerBuildVersion = findViewById(R.id.header_build_version);
        headerBuildVersion.setText(
                getString(R.string.header_android_version, Build.VERSION.RELEASE));

        TextView headerBuildDate = findViewById(R.id.header_build_date);
        headerBuildDate.setText(StringGenerator.getDateLocalizedUTC(this,
                DateFormat.LONG, DeviceInfoUtils.getBuildDateTimestamp()));

        TextView headerSecurityPatch = findViewById(R.id.header_security_patch_level);
        headerSecurityPatch.setText(getString(R.string.header_android_security_update,
                DeviceInfoUtils.getSecurityPatch()));

        maybeShowWelcomeMessage();
    }

    @Override
    public void onStart() {
        super.onStart();
        Intent intent = new Intent(this, UpdaterService.class);
        startService(intent);
        bindService(intent, mConnection, Context.BIND_AUTO_CREATE);

        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(UpdaterController.ACTION_UPDATE_STATUS);
        intentFilter.addAction(UpdaterController.ACTION_DOWNLOAD_PROGRESS);
        intentFilter.addAction(UpdaterController.ACTION_INSTALL_PROGRESS);
        intentFilter.addAction(UpdaterController.ACTION_UPDATE_REMOVED);
        LocalBroadcastManager.getInstance(this).registerReceiver(mBroadcastReceiver, intentFilter);
    }

    @Override
    protected void onPause() {
        if (importDialog != null) {
            importDialog.dismiss();
            importDialog = null;
            mUpdateImporter.stopImport();
        }

        super.onPause();
    }

    @Override
    public void onStop() {
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mBroadcastReceiver);
        if (mUpdaterService != null) {
            unbindService(mConnection);
        }
        super.onStop();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (!mUpdateImporter.onResult(requestCode, resultCode, data)) {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    @Override
    public void onRefreshClick() {
        mViewModel.refreshUpdates();
    }

    @Override
    public void onLocalUpdateClick() {
        mUpdateImporter.openImportPicker();
    }

    @Override
    public void onPreferencesClick() {
        showPreferencesDialog();
    }

    @Override
    public void onChangelogClick() {
        Intent openUrl = new Intent(Intent.ACTION_VIEW,
                Uri.parse(Utils.getChangelogURL(this)));
        startActivity(openUrl);
    }

    @Override
    public void onImportStarted() {
        if (importDialog != null && importDialog.isShowing()) {
            importDialog.dismiss();
        }

        importDialog = new AlertDialog.Builder(this)
                .setTitle(R.string.local_update_import)
                .setView(R.layout.progress_dialog)
                .setCancelable(false)
                .create();

        importDialog.show();
    }

    @Override
    public void onImportCompleted(Update update) {
        if (importDialog != null) {
            importDialog.dismiss();
            importDialog = null;
        }

        if (update == null) {
            new AlertDialog.Builder(this)
                    .setTitle(R.string.local_update_import)
                    .setMessage(R.string.local_update_import_failure)
                    .setPositiveButton(android.R.string.ok, null)
                    .show();
            return;
        }

        mAdapter.notifyDataSetChanged();

        final Runnable deleteUpdate = () -> UpdaterController.getInstance(this)
                .deleteUpdate(update.getDownloadId());

        new AlertDialog.Builder(this)
                .setTitle(R.string.local_update_import)
                .setMessage(getString(R.string.local_update_import_success, update.getVersion()))
                .setPositiveButton(R.string.local_update_import_install, (dialog, which) -> {
                    mAdapter.addItem(update.getDownloadId());
                    mViewModel.refreshUpdates();
                    Utils.triggerUpdate(this, update.getDownloadId());
                })
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> deleteUpdate.run())
                .setOnCancelListener((dialog) -> deleteUpdate.run())
                .show();
    }

    private final ServiceConnection mConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName className,
                IBinder service) {
            UpdaterService.LocalBinder binder = (UpdaterService.LocalBinder) service;
            mUpdaterService = binder.getService();
            UpdaterController controller = mUpdaterService.getUpdaterController();
            mAdapter.setUpdaterController(controller);
            mViewModel.refreshUpdates();
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            mAdapter.setUpdaterController(null);
            mUpdaterService = null;
            mAdapter.notifyDataSetChanged();
        }
    };

    private void updateLastCheckedString(long lastCheckMillis) {
        if (lastCheckMillis <= 0) return;
        long lastCheckSeconds = lastCheckMillis / 1000;
        String lastCheckString = getString(R.string.header_last_updates_check,
                StringGenerator.getDateLocalized(this, DateFormat.LONG, lastCheckSeconds),
                StringGenerator.getTimeLocalized(this, lastCheckSeconds));
        ((TextView) findViewById(R.id.header_last_check)).setText(lastCheckString);
    }

    private void handleDownloadStatusChange(String downloadId) {
        if (Update.LOCAL_ID.equals(downloadId)) {
            return;
        }

        Update update = mUpdaterService.getUpdaterController().getUpdate(downloadId);
        switch (update.getStatus()) {
            case PAUSED_ERROR:
                showToast(R.string.snack_download_failed, Toast.LENGTH_LONG);
                break;
            case VERIFICATION_FAILED:
                showToast(R.string.snack_download_verification_failed, Toast.LENGTH_LONG);
                break;
            case VERIFIED:
                showToast(R.string.snack_download_verified, Toast.LENGTH_LONG);
                break;
        }
    }

    private void exportUpdate(Update update) {
        mToBeExported = update;

        Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("application/zip");
        intent.putExtra(Intent.EXTRA_TITLE, update.getName());

        mExportUpdate.launch(intent);
    }

    private void exportUpdate(Uri uri) {
        Intent intent = new Intent(this, ExportUpdateService.class);
        intent.setAction(ExportUpdateService.ACTION_START_EXPORTING);
        intent.putExtra(ExportUpdateService.EXTRA_SOURCE_FILE, mToBeExported.getFile());
        intent.putExtra(ExportUpdateService.EXTRA_DEST_URI, uri);
        startService(intent);
    }

    private void showToast(int stringId, int duration) {
        Toast.makeText(this, stringId, duration).show();
    }

    private void showPreferencesDialog() {
        View view = LayoutInflater.from(this).inflate(R.layout.preferences_dialog, null);
        Spinner autoCheckInterval = view.findViewById(R.id.preferences_auto_updates_check_interval);
        Switch autoDelete = view.findViewById(R.id.preferences_auto_delete_updates);
        Switch meteredNetworkWarning = view.findViewById(
                R.id.preferences_metered_network_warning);
        Switch abPerfMode = view.findViewById(R.id.preferences_ab_perf_mode);
        Switch updateRecovery = view.findViewById(R.id.preferences_update_recovery);

        if (!DeviceInfoUtils.isABDevice()) {
            abPerfMode.setVisibility(View.GONE);
        }

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        autoCheckInterval.setSelection(Utils.getUpdateCheckSetting(this));
        autoDelete.setChecked(prefs.getBoolean(Constants.PREF_AUTO_DELETE_UPDATES, false));
        meteredNetworkWarning.setChecked(prefs.getBoolean(Constants.PREF_METERED_NETWORK_WARNING, true));
        abPerfMode.setChecked(prefs.getBoolean(Constants.PREF_AB_PERF_MODE, false));
        boolean periodicCheckEnabled = prefs.getBoolean(
                Constants.PREF_PERIODIC_CHECK_ENABLED, true);
        String periodicCheckIntervalValue = prefs.getString(
                CheckInterval.PREF_KEY, CheckInterval.WEEKLY.getValue());

        // Only show update recovery option for non-AB devices with recovery update script
        if (Utils.isRecoveryUpdateExecPresent()) {
            updateRecovery.setVisibility(View.VISIBLE);
            // Obtain and apply the user preference from SetupWizard.
            updateRecovery.setChecked(DeviceInfoUtils.isRecoveryUpdateEnabled());
        } // else: remains GONE (default from XML)

        new AlertDialog.Builder(this)
                .setTitle(R.string.menu_preferences)
                .setView(view)
                .setOnDismissListener(dialogInterface -> {
                    int intervalPos = autoCheckInterval.getSelectedItemPosition();
                    boolean checkEnabled = intervalPos != 0;
                    String intervalValue = switch (intervalPos) {
                        case 1 -> CheckInterval.DAILY.getValue();
                        case 3 -> CheckInterval.MONTHLY.getValue();
                        default -> CheckInterval.WEEKLY.getValue();
                    };
                    boolean scheduleChanged = periodicCheckEnabled != checkEnabled
                            || (checkEnabled && !intervalValue.equals(periodicCheckIntervalValue));

                    prefs.edit()
                            .putBoolean(Constants.PREF_PERIODIC_CHECK_ENABLED, checkEnabled)
                            .putString(CheckInterval.PREF_KEY, intervalValue)
                            .putBoolean(Constants.PREF_AUTO_DELETE_UPDATES, autoDelete.isChecked())
                            .putBoolean(Constants.PREF_METERED_NETWORK_WARNING,
                                    meteredNetworkWarning.isChecked())
                            .putBoolean(Constants.PREF_AB_PERF_MODE, abPerfMode.isChecked())
                            .apply();

                    if (scheduleChanged) {
                        if (checkEnabled) {
                            UpdateCheckWorker.reschedulePeriodicCheck(this);
                        } else {
                            UpdateCheckWorker.cancelPeriodicCheck(this);
                        }
                    }

                    if (DeviceInfoUtils.isABDevice()) {
                        boolean enableABPerfMode = abPerfMode.isChecked();
                        mUpdaterService.getUpdaterController().setPerformanceMode(enableABPerfMode);
                    }
                    if (Utils.isRecoveryUpdateExecPresent()) {
                        boolean enableRecoveryUpdate = updateRecovery.isChecked();
                        DeviceInfoUtils.setRecoveryUpdateEnabled(enableRecoveryUpdate);
                    }
                })
                .show();
    }

    private void maybeShowWelcomeMessage() {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        boolean alreadySeen = preferences.getBoolean(Constants.PREF_HAS_SEEN_WELCOME_MESSAGE, false);
        if (alreadySeen) {
            return;
        }
        new AlertDialog.Builder(this)
                .setTitle(R.string.welcome_title)
                .setMessage(R.string.welcome_message)
                .setPositiveButton(R.string.info_dialog_ok, (dialog, which) -> preferences.edit()
                        .putBoolean(Constants.PREF_HAS_SEEN_WELCOME_MESSAGE, true)
                        .apply())
                .show();
    }
}

File: app/src/main/java/org/lineageos/updater/UpdatesActivity.kt
/*
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.updater

import android.app.AlertDialog as ViewAlertDialog
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.ServiceConnection
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.adaptive.currentWindowAdaptiveInfo
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.DialogProperties
import androidx.window.core.layout.WindowSizeClass
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import com.android.settingslib.spa.framework.compose.LifecycleEffect
import com.android.settingslib.spa.framework.compose.LocalNavController
import com.android.settingslib.spa.framework.compose.NavControllerWrapper
import com.android.settingslib.spa.framework.theme.SettingsTheme
import com.android.settingslib.spa.framework.util.collectLatestWithLifecycle
import com.android.settingslib.spa.widget.preference.Preference
import com.android.settingslib.spa.widget.preference.PreferenceModel
import com.android.settingslib.spa.widget.scaffold.RegularScaffold
import com.android.settingslib.spa.widget.scaffold.SettingsScaffold
import com.android.settingslib.spa.widget.ui.Category
import com.android.settingslib.spa.widget.ui.LinearLoadingBar
import org.lineageos.updater.controller.UpdaterController
import org.lineageos.updater.controller.UpdaterService
import org.lineageos.updater.data.Update
import org.lineageos.updater.data.UpdateStatus
import org.lineageos.updater.misc.Utils
import org.lineageos.updater.preferences.PreferencesActivity
import org.lineageos.updater.update.UpdateList
import org.lineageos.updater.util.NetworkMonitor

class UpdatesActivity : ComponentActivity(), UpdateImporter.Callbacks {

    private var updaterService: UpdaterService? = null
    private var isImporting by mutableStateOf(false)
    private var toBeExported: Update? = null

    private val viewModel: UpdatesViewModel by viewModels { UpdatesViewModel.factory(application) }
    private val updateImporter = UpdateImporter(this, this)
    private val updaterController = mutableStateOf<UpdaterController?>(null)
    private val progressRevision = mutableIntStateOf(0)

    private val exportResultLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            result.data?.data?.let { uri -> performExport(uri) }
        }
    }

    private val broadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.action) {
                UpdaterController.ACTION_UPDATE_STATUS -> {
                    val downloadId = intent.getStringExtra(UpdaterController.EXTRA_DOWNLOAD_ID)
                    handleDownloadStatusChange(downloadId)
                    progressRevision.intValue++
                }

                UpdaterController.ACTION_DOWNLOAD_PROGRESS,
                UpdaterController.ACTION_INSTALL_PROGRESS,
                UpdaterController.ACTION_UPDATE_REMOVED -> progressRevision.intValue++
            }
        }
    }

    private val connection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            val updaterService = (service as UpdaterService.LocalBinder).service
            this@UpdatesActivity.updaterService = updaterService
            updaterController.value = updaterService.updaterController
            refreshUpdatesList(viewModel.uiState.value.updates)
        }

        override fun onServiceDisconnected(componentName: ComponentName) {
            updaterController.value = null
            updaterService = null
            progressRevision.intValue++
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        setContent {
            val navController = remember {
                object : NavControllerWrapper {
                    override fun navigate(route: String, popUpCurrent: Boolean) {}
                    override fun navigateBack() = finish()
                }
            }
            CompositionLocalProvider(LocalNavController provides navController) {
                SettingsTheme {
                    LifecycleEffect(
                        onStart = {
                            val intent = Intent(this@UpdatesActivity, UpdaterService::class.java)
                            startService(intent)
                            bindService(intent, connection, BIND_AUTO_CREATE)
                            registerReceiver(
                                broadcastReceiver,
                                IntentFilter().apply {
                                    addAction(UpdaterController.ACTION_UPDATE_STATUS)
                                    addAction(UpdaterController.ACTION_DOWNLOAD_PROGRESS)
                                    addAction(UpdaterController.ACTION_INSTALL_PROGRESS)
                                    addAction(UpdaterController.ACTION_UPDATE_REMOVED)
                                },
                                RECEIVER_NOT_EXPORTED,
                            )
                        },
                        onStop = {
                            unregisterReceiver(broadcastReceiver)
                            if (updaterService != null) unbindService(connection)
                        },
                    )
                    if (isImporting) {
                        AlertDialog(
                            onDismissRequest = {},
                            properties = DialogProperties(
                                dismissOnBackPress = false,
                                dismissOnClickOutside = false,
                            ),
                            title = { Text(stringResource(R.string.local_update_import)) },
                            text = {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    CircularProgressIndicator(modifier = Modifier.size(40.dp))
                                    Spacer(Modifier.width(16.dp))
                                    Text(stringResource(R.string.local_update_import_progress))
                                }
                            },
                            confirmButton = {},
                        )
                    }
                    val uiState by viewModel.uiState.collectAsState()
                    val networkState by NetworkMonitor.getInstance(this)
                        .networkState.collectAsState()
                    val updates = uiState.updates
                    val titleText = when {
                        updates.isEmpty() && uiState.lastCheckedTimestamp == 0L ->
                            stringResource(R.string.display_name)

                        updates.isEmpty() -> stringResource(R.string.snack_no_updates_found)
                        updates.any { it.status == UpdateStatus.UPDATED_NEED_REBOOT } ->
                            stringResource(R.string.installing_update_finished)

                        updates.any {
                            it.status == UpdateStatus.INSTALLING ||
                                    it.status == UpdateStatus.INSTALLATION_SUSPENDED
                        } -> stringResource(R.string.installing_update)

                        updates.any { it.status == UpdateStatus.VERIFYING } ->
                            stringResource(R.string.verifying_download_notification)

                        updates.any {
                            it.status == UpdateStatus.DOWNLOADING ||
                                    it.status == UpdateStatus.STARTING ||
                                    it.status == UpdateStatus.PAUSED ||
                                    it.status == UpdateStatus.PAUSED_ERROR
                        } -> stringResource(R.string.downloading_notification)

                        else -> stringResource(R.string.snack_updates_found)
                    }
                    val windowAdaptiveInfo = currentWindowAdaptiveInfo()
                    val isWideScreen = windowAdaptiveInfo.windowSizeClass
                        .isWidthAtLeastBreakpoint(WindowSizeClass.WIDTH_DP_MEDIUM_LOWER_BOUND)
                    val localUpdateSummary = stringResource(R.string.local_update_import_summary)
                    val preferencesSummary = stringResource(R.string.preferences_summary)

                    val footerPane: @Composable () -> Unit = {
                        Category {
                            Preference(object : PreferenceModel {
                                override val title = stringResource(R.string.local_update_import)
                                override val summary = { localUpdateSummary }
                                override val onClick = { updateImporter.openImportPicker() }
                            })
                            Preference(object : PreferenceModel {
                                override val title = stringResource(R.string.menu_preferences)
                                override val summary = { preferencesSummary }
                                override val onClick = {
                                    startActivity(
                                        Intent(
                                            this@UpdatesActivity,
                                            PreferencesActivity::class.java,
                                        )
                                    )
                                }
                            })
                        }
                    }

                    if (isWideScreen) {
                        SettingsScaffold(title = titleText) { paddingValues ->
                            Column(modifier = Modifier.fillMaxSize()) {
                                Spacer(Modifier.height(paddingValues.calculateTopPadding()))
                                LinearLoadingBar(isLoading = uiState.isCheckingForUpdates)
                                Row(modifier = Modifier
                                    .weight(1f)
                                    .fillMaxWidth()) {
                                    Column(
                                        modifier = Modifier
                                            .weight(1f)
                                            .fillMaxHeight()
                                            .verticalScroll(rememberScrollState()),
                                    ) {
                                        DeviceInfoBanner()
                                        Spacer(Modifier.height(paddingValues.calculateBottomPadding()))
                                    }
                                    Column(
                                        modifier = Modifier
                                            .weight(1f)
                                            .fillMaxHeight()
                                            .verticalScroll(rememberScrollState()),
                                    ) {
                                        UpdatesCheck(
                                            isRefreshing = uiState.isCheckingForUpdates,
                                            isNetworkAvailable = networkState.isOnline,
                                            lastCheckTimestamp = uiState.lastCheckedTimestamp,
                                            onCheckClick = viewModel::fetchUpdates,
                                        )
                                        UpdateList(
                                            updates = updates,
                                            updaterController = updaterController.value,
                                            networkState = networkState,
                                            onExportUpdate = ::onExportUpdate,
                                            progressRevision = progressRevision.intValue,
                                        )
                                        footerPane()
                                        Spacer(Modifier.height(paddingValues.calculateBottomPadding()))
                                    }
                                }
                            }
                        }
                    } else {
                        RegularScaffold(title = titleText) {
                            LinearLoadingBar(isLoading = uiState.isCheckingForUpdates)
                            DeviceInfoBanner()
                            UpdatesCheck(
                                isRefreshing = uiState.isCheckingForUpdates,
                                isNetworkAvailable = networkState.isOnline,
                                lastCheckTimestamp = uiState.lastCheckedTimestamp,
                                onCheckClick = viewModel::fetchUpdates,
                            )
                            UpdateList(
                                updates = updates,
                                updaterController = updaterController.value,
                                networkState = networkState,
                                onExportUpdate = ::onExportUpdate,
                                progressRevision = progressRevision.intValue,
                            )
                            footerPane()
                        }
                    }
                }
            }
        }

        viewModel.uiState.collectLatestWithLifecycle(this) { state ->
            if (state.errorMessage != null) {
                Toast.makeText(this, R.string.snack_updates_check_failed, Toast.LENGTH_LONG).show()
                viewModel.errorMessageShown()
            }
            if (updaterService != null) {
                refreshUpdatesList(state.updates)
            }
        }
    }

    override fun onPause() {
        if (isImporting) {
            isImporting = false
            updateImporter.stopImport()
        }
        super.onPause()
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (!updateImporter.onResult(requestCode, resultCode, data)) {
            @Suppress("DEPRECATION")
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onImportStarted() {
        isImporting = true
    }

    override fun onImportCompleted(update: Update?) {
        isImporting = false

        if (update == null) {
            ViewAlertDialog.Builder(this)
                .setTitle(R.string.local_update_import)
                .setMessage(R.string.local_update_import_failure)
                .setPositiveButton(android.R.string.ok, null)
                .show()
            return
        }

        val deleteUpdate = { UpdaterController.getInstance(this).deleteUpdate(update.downloadId) }

        ViewAlertDialog.Builder(this)
            .setTitle(R.string.local_update_import)
            .setMessage(getString(R.string.local_update_import_success, update.version))
            .setPositiveButton(R.string.local_update_import_install) { _, _ ->
                refreshUpdatesList(viewModel.uiState.value.updates)
                Utils.triggerUpdate(this, update.downloadId)
            }
            .setNegativeButton(android.R.string.cancel) { _, _ -> deleteUpdate() }
            .setOnCancelListener { deleteUpdate() }
            .show()
    }

    private fun refreshUpdatesList(updates: List<Update>) {
        val controller = updaterService?.updaterController ?: return
        val updateIds = updates.map { update ->
            controller.addUpdate(update)
            update.downloadId
        }
        controller.setUpdatesAvailableOnline(updateIds, true)
    }

    private fun handleDownloadStatusChange(downloadId: String?) {
        if (downloadId == Update.LOCAL_ID) return
        val update = updaterService?.updaterController?.getUpdate(downloadId) ?: return
        when (update.status) {
            UpdateStatus.PAUSED_ERROR ->
                Toast.makeText(this, R.string.snack_download_failed, Toast.LENGTH_LONG).show()

            UpdateStatus.VERIFICATION_FAILED ->
                Toast.makeText(
                    this, R.string.snack_download_verification_failed, Toast.LENGTH_LONG,
                ).show()

            UpdateStatus.VERIFIED ->
                Toast.makeText(this, R.string.snack_download_verified, Toast.LENGTH_LONG).show()

            else -> {}
        }
    }

    private fun onExportUpdate(update: Update) {
        toBeExported = update
        exportResultLauncher.launch(
            Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
                addCategory(Intent.CATEGORY_OPENABLE)
                type = "application/zip"
                putExtra(Intent.EXTRA_TITLE, update.name)
            }
        )
    }

    private fun performExport(uri: Uri) {
        val file = toBeExported?.file ?: return
        startService(
            Intent(this, ExportUpdateService::class.java).apply {
                action = ExportUpdateService.ACTION_START_EXPORTING
                putExtra(ExportUpdateService.EXTRA_SOURCE_FILE, file)
                putExtra(ExportUpdateService.EXTRA_DEST_URI, uri)
            }
        )
    }
}
